feat(PinInput): implement component

maxsteinwand opened the pull request:
### 🔗 Linked issue

Resolves #1123 

### ❓ Type of change
- [X] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality)
- [X] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description
Add the PinInput component from radix-vue. https://www.radix-vue.com/components/pin-input
Tests and Documentation are also added.

### 📝 Checklist
- [X] I have linked an issue or discussion.
- [X] I have updated the documentation accordingly.

Commit:
test

maxsteinwand committed:
Merge remote-tracking branch 'origin/v3' into PinInput

maxsteinwand committed:
docs + style

maxsteinwand committed:
docs

maxsteinwand committed:
renaming

maxsteinwand committed:
RootEmits

maxsteinwand committed:
Rename PinInput.vue to Pininput.vue

maxsteinwand committed:
renaming

maxsteinwand committed:
Rename Pininput.vue to PinInput.vue

maxsteinwand committed:
tests and more docs

maxsteinwand committed:
Merge branch 'PinInput' of https://github.com/maxsteinwand/ui into PinInput

maxsteinwand committed:
reorder router import

maxsteinwand committed:
Merge branch 'v3' into PinInput

maxsteinwand commented:
I hope this implementation is okay even tough the radix-vue component is tagged as alpha in their docs. It will not work in a Form (validation etc.) i struggled to get that working. Hope the tests are good/fine as well as i have no expierence with that so far 😄 
I oriented myself on the normal input and some other radix-vue implementations.

benjamincanac commented on docs/content/3.components/pin-input.md:
```diff
@@ -0,0 +1,166 @@
+---
+::
+
+### Default Value
```
I usually mention the `default-value` in the first block: https://ui3.nuxt.dev/components/switch#usage

maxsteinwand replied:
I can change that no problem

benjamincanac commented on docs/content/3.components/pin-input.md:
```diff
@@ -0,0 +1,166 @@
+---
+::
+
+### Mask
```
Could it be worth having an example with `otp` prop? Not sure what it does exactly though.

maxsteinwand replied:
I was thinking about that too but there is nothing to showcase. It does set the autocomplete type to "one-time-code". This is defined here: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete#one-time-code
which then allows mobile users to autocomplete with the clipboard or from recieved messages.
https://github.com/unovue/radix-vue/blob/main/packages/radix-vue/src/PinInput/PinInputInput.vue#L159

benjamincanac replied:
~~Shouldn't this be set by default then? 🤔~~ Nevermind, I added it to the docs!

benjamincanac commented on src/runtime/components/PinInput.vue:
```diff
@@ -0,0 +1,65 @@
+}
+
+export interface PinInputEmits extends PinInputRootEmits {}
+export interface PinInputSlots {}
```
No need to define slots if there are none I guess.

benjamincanac commented on src/runtime/components/PinInput.vue:
```diff
@@ -0,0 +1,65 @@
+  type: 'text',
+  length: 5
+})
+// const slots = defineSlots<PinInputSlots>()
```
Same here

benjamincanac commented on src/theme/pin-input.ts:
```diff
@@ -0,0 +1,69 @@
+
+export default (options: Required<ModuleOptions>) => ({
+  slots: {
+    root: 'flex gap-2 items-center mt-1',
```
Why add a `mt-1` here?

maxsteinwand replied:
I'll remove it, was copy paste classes from the radix-vue example 😁

benjamincanac reviewed (commented):
Wow 🤩 Thank you for doing this, you just saved me a lot of time and the PR looks perfect! 

Did you find the CLI to start the component?

benjamincanac commented:
You also have to run `pnpm test:vue` to create the snapshots 😊 I haven't found a way to execute both at the same time yet.

benjamincanac commented:
@romhml Could you have a look into this to handle it inside a form?

maxsteinwand committed:
cleanup

maxsteinwand committed:
Merge branch 'PinInput' of https://github.com/maxsteinwand/ui into PinInput

maxsteinwand committed:
rerun tests

maxsteinwand committed:
Merge branch 'v3' into PinInput

maxsteinwand commented:
@benjamincanac Thanks for the quick feedback! I adjusted the docs as you said. Yeah i found the cli, but actually after i already created the files and then i renamed everything again😄

maxsteinwand committed:
fix test spec

maxsteinwand committed:
Merge branch 'PinInput' of https://github.com/maxsteinwand/ui into PinInput

benjamincanac committed:
Merge branch 'v3' into PinInput

benjamincanac committed:
playground(pin-input): update

benjamincanac committed:
fix(PinInput): clean props

benjamincanac committed:
improve theme (same size as inputs)

benjamincanac committed:
test: update

benjamincanac committed:
docs(ComponentCode): fix numeric values

benjamincanac committed:
docs(pin-input): update

benjamincanac committed:
test: update snapshots

maxsteinwand committed:
Merge branch 'v3' into PinInput

romhml committed:
feat(PinInput): integrate with Form component

romhml commented:
I've made some changes to integrate the PinInput with the Form/FormField components. One thing I could not figure out is how to emit a `@blur`  event, @maxsteinwand do you have an idea?

maxsteinwand committed:
feat(PinInput): add blur event

maxsteinwand committed:
update imports

maxsteinwand commented:
@romhml i added the blur event. The event.relatedTarget will be null if the user leaves the input. It would be set to the next input element if the user enters a value. But i am still thinking about the change event. If you now test on the playground and you enter all 5 values it will validate and mark the field as invalid (because max. 4 allowed) but if you then remove the last entry it will not validate again as the change will only get called once all fields are filled.

romhml commented:
I made the playground example to test that the validation is triggered, I'll update it to have something more realistic.

romhml committed:
fix(PinInput): add form validation on blur

romhml commented:
I added the form validation on blur and updated the example. The playground example does not work properly currently because the input uses an array rule, I'll do another PR to fix this.

romhml commented:
I added the ` error-pattern`  prop [here](https://github.com/nuxt/ui/pull/2601) which should be used in the PinInput example to correctly associate the errors with the form field.

romhml committed:
Merge branch 'v3' of github.com:nuxt/ui into PinInput

benjamincanac committed:
Merge branch 'v3' into PinInput

romhml committed:
chore(Form): update PinInput form example

romhml committed:
Merge branch 'v3' of github.com:nuxt/ui into PinInput

benjamincanac committed:
Merge branch 'v3' into PinInput

benjamincanac committed:
fix(PinInput): add missing `ref`

benjamincanac committed:
test: update

benjamincanac committed:
fix(Input/PinInput): add missing `bg-transparent`

benjamincanac committed:
fix(Input): put back gaps

benjamincanac committed:
docs(FormExampleElements): add w-full

benjamincanac committed:
test: update snapshots

benjamincanac committed:
up

benjamincanac commented:
Thanks @maxsteinwand! 😊
